File: 2024/elixir/day04.exs
# https://adventofcode.com/2024/day/4

defmodule Day04 do
  @dirs [{1, 0}, {0, 1}, {0, -1}, {-1, 0}, {1, 1}, {1, -1}, {-1, 1}, {-1, -1}]

  def part1(input) do
    grid = to_grid(input)

    for {point, _} <- grid, reduce: 0 do
      total ->
        total + Enum.count(words_for_point(grid, point), &(&1 == "XMAS"))
    end
  end

  def part2(input) do
    grid = to_grid(input)

    for {point, c} <- grid, c == "A", reduce: 0 do
      total ->
        if xmas?(grid, point), do: total + 1, else: total
    end
  end

  defp xmas?(grid, {x, y}) do
    d1 = Enum.sort([grid[{x - 1, y - 1}], grid[{x, y}], grid[{x + 1, y + 1}]])
    d2 = Enum.sort([grid[{x - 1, y + 1}], grid[{x, y}], grid[{x + 1, y - 1}]])

    d1 == ~w[A M S] and d2 == ~w[A M S]
  end

  defp words_for_point(grid, {x, y}) do
    Enum.map(@dirs, fn {dir_x, dir_y} ->
      Enum.reduce(0..3, "", fn delta, word ->
        point = {x + delta * dir_x, y + delta * dir_y}

        case grid[point] do
          nil -> word
          _ -> word <> grid[point]
        end
      end)
    end)
  end

  defp to_grid(input) do
    input
    |> String.split("\n", trim: true)
    |> then(fn rows ->
      for {row, y} <- Enum.with_index(rows),
          {cell, x} <- Enum.with_index(String.split(row, "", trim: true)),
          into: %{},
          do: {{x, y}, cell}
    end)
  end
end

"""
MMMSXXMASM
MSAMXMSMSA
AMXSXMAAMM
MSAMASMSMX
XMASAMXAMM
XXAMMXXAMA
SMSMSASXSS
SAXAMASAAA
MAMMMXMMMM
MXMXAXMASX
"""
|> then(fn input ->
  {
    Day04.part1(input),
    Day04.part2(input)
  }
  |> IO.inspect()
end)

File: 2024/ruby/day04.rb
#!/usr/bin/env ruby

require_relative "grid"

GRID = Grid.parse(DATA)
DIRS = [[1, 0], [0, 1], [0, -1], [-1, 0], [1, 1], [1, -1], [-1, 1], [-1, -1]]

def xmas_count(point)
  return 0 unless GRID.at(point) == "X"
  count = 0

  DIRS.each { |(dx, dy)|
    (0..3).reduce("") { |word, delta|
      if (letter = GRID.at(point.add(delta * dx, delta * dy)))
        word + letter
      else
        word
      end
    }.tap { count += 1 if _1 == "XMAS" }
  }

  count
end

def x_mas?(point)
  return unless GRID.at(point) == "A"

  d1 = [GRID.at(point.x - 1, point.y - 1), GRID.at(point.x + 1, point.y + 1)].compact.sort
  d2 = [GRID.at(point.x - 1, point.y + 1), GRID.at(point.x + 1, point.y - 1)].compact.sort

  d1 == %w[M S] && d2 == %w[M S]
end

p1 = p2 = 0

GRID.each_point { |point|
  p1 += xmas_count(point)
  p2 += 1 if x_mas?(point)
}

puts "P1: #{p1}"
puts "P2: #{p2}"

__END__
MMMSXXMASM
MSAMXMSMSA
AMXSXMAAMM
MSAMASMSMX
XMASAMXAMM
XXAMMXXAMA
SMSMSASXSS
SAXAMASAAA
MAMMMXMMMM
MXMXAXMASX
